### PR TITLE
Feature/create habit structure

### PR DESCRIPTION
**Pull request by kmaaallen**
Contained in this PR:

New:
- CreateHabit component
- createHabit action
- Tests for the above

Updated:
- Count reducer - to reflect new habit structure and handle create habit action
- Selector - updated as passing state.count from logbutton
- Log page- to contain create habit component and all habits for user
- App - to contain react native paper provider at top level for use of 'Portal' at lower levels
- LogButton component - to contain Portal for dialog display and to display content based on passed habit id

Comments:
- Had to mock react native paper Provider component as PaperProvider in App test to ensure CI tests pass though was still passing locally. Hence a lot of commits around testing as working this out.
- LogButton mapStateToProps feels clunky - some refactoring probably needed

### CODE REVIEW

File: src/components/CreateHabit.js
Comment (jdbann) [inline src/components/CreateHabit.js:12]: A few things here on the styling. Is it possible for the widths be percentages? My iOS Simulator defaults to an iPhone 12 Pro Max and these elements are left-aligned and only use about 80% of the screen.

Additionally, it looks like some of the margins are different between the the `CreateHabit` component and the `LogButton` component. I think it's just that one is 20px and one is 25px, but two options pop to mind to ensure this is consistent in future:
- Setup a `styles` object which includes various spacing sizes and pull values from there (so something like `margin: styles.spacing.medium`)
- Wrap the elements in a component and apply padding to that parent component instead of applying margin to the children components (except for whatever margins you need to create space between the children components)

File: src/components/CreateHabit.js
Comment (jdbann) [inline src/components/CreateHabit.js]: You could avoid repeating a lot of boilerplate stuff here with hooks:

```js
export const CreateHabit = () => {
  const [showDialoag, setShowDialog] = useState(false)

  return (
    <Button onPress={() => this.setShowDialog(true)}> Create new habit</Button>
  )
}
```

- [React Hooks](https://reactjs.org/docs/hooks-intro.html)
- [Redux Hooks](https://react-redux.js.org/api/hooks)

File: src/components/LogButton.js
Comment (jdbann) [inline src/components/LogButton.js:72]: I don't think we can get this case in the app with the way it's put together right now, but if `habit` was `null` then `lastIncrementLog` would also be `null` - so here we'd get `Uncaught TypeError: Cannot read property '0' of null`

If `lastIncrementLog` equalled `[]` when `habit` was `null` I think we'd be safe

File: src/redux/actions.js
Comment (jdbann) [inline src/redux/actions.js]: I don't know where I got this from, but when I use ordered params, I have a tendency to order my method params in the order they would be used:

- Values I need in order to find the right object
- Values I need in order to perform the right action on the object
- Values I might not need or use

This is building an action so they all get used at the same time, but they essentially call the reducer so my head still turns to this. So I'd but `id` here first... Or using named params avoids all this too...

File: src/redux/reducers.js
Comment (jdbann) [inline src/redux/reducers.js]: This is probably a `habitReducer` now

File: src/redux/reducers.js
Comment (jdbann) [inline src/redux/reducers.js]: I think these actions are making modifications to the original state value. You do use the spread operator in the return which will create a clone of the state rather than the same state object - but the original state object is updated with the updated habit details before this.

The spread operator makes a shallow clone, so when you spread an object the values are still pointing to the initial object's value. Here's an example of what I mean that I just ran in the `node` console:

```js
// Setup an initial state
state = { habit: { 0: { name: "Take meds" }, 1: { name: "Exercise" } } }

// Modify a habit
state.habit[0].name = "New name"

// Spread the initial state into a new state object - the new habit name is in place
newState = { ...state }
// > { habit: { '0': { name: 'New name' }, '1': { name: 'Exercise' } } }

// Inspect the original state - the new name is also in place
state
// > { habit: { '0': { name: 'New name' }, '1': { name: 'Exercise' } } }
```

Redux needs to be careful about this for a few reasons, simplest of which is that it means the library can no longer compare the old state to the new state in order to see which components will need to re-render.

- [Reducers and Immutable Updates](https://redux.js.org/tutorials/fundamentals/part-3-state-actions-reducers#reducers-and-immutable-updates)
- [Immutable Update Patterns](https://redux.js.org/usage/structuring-reducers/immutable-update-patterns#simplifying-immutable-updates-with-redux-toolkit)

File: tests/App.test.js
Comment (jdbann) [inline tests/App.test.js:12]: Nice 👍

File: tests/CreateHabit.test.js
Comment (jdbann) [inline tests/CreateHabit.test.js:53]: With `react-test-renderer`, can you get the props and call actual functions? Would we be able to test that the store is modified if we complete the modal and submit?

File: tests/LogButton.test.js
Comment (jdbann) [inline tests/LogButton.test.js:41]: Can I ask why we need to mock out the `toLocaleString` method here?

**Review by jdbann — APPROVED** [review]
Firstly, I want to apologise for taking so long to get around to reviewing this again. Been a weird couple of weeks but...

Secondly, this was a joy to read - exactly what I wanted from a Sunday morning and even though that sounds sarcastic I mean it. I've been picky in some places but I think they'll be interesting areas to get into.

I'm aware this PR has been open for a while though. The tests pass and the app performs as described in the simulator, so feel free to respond to comments, merge what you have, and create issues for anything you think is worth investigating further.

Regarding your comments:
- Mocking the `PortalManager` was a canny move! Something like that, where it's not immediately obvious why something is done in the code, is probably a worthwhile place for a comment in the code itself - saves debugging in future when it is accidentally deleted!
- `mapStateToProps` functions can get convoluted quickly. Looking at hooks will help with that, but we could also be more explicit about using `mapStateToProps` and `mapDispatchToProps` for getting and setting values - but do all formatting of those values in the function directly.

**Comment by jdbann** [comment]
Oh we could also talk a bit about 'git hygiene' (although I hate that term for these kind of things as it's very opinionated and judgemental).

The main thing is that git can be a really powerful tool for adding context to the code, in addition to documentation. Some definitions:

- **Documentation** - Explanations of what the code does, how it does it and how to use it safely
- **Context** - How the code has ended up working this way, changes that have been required, or limitations which have shaped the solution

So a really useful feature that git offers is `git blame`. It can show you which commit is responsible for each change in a file. Here's the `tests/App.test.js` file:

<details>
<summary>$ git blame -s tests/App.test.js</summary>

```
$ git blame -s tests/App.test.js
8477bd2e App.test.js        1) import React from 'react';
8477bd2e App.test.js        2) import renderer from 'react-test-renderer';
3e8b3951 tests/App.test.js  3) import App from '../App';
8477bd2e App.test.js        4)
3e8b3951 tests/App.test.js  5) jest.mock('redux-persist/integration/react', () => ({
3e8b3951 tests/App.test.js  6)     PersistGate: props => props.children,
3e8b3951 tests/App.test.js  7) }));
8477bd2e App.test.js        8)
5b512532 tests/App.test.js  9) jest.mock('react-native-paper/lib/commonjs/core/Provider', () => 'PaperProvider');
5b512532 tests/App.test.js 10)
8477bd2e App.test.js       11) describe('<App />', () => {
86fc8623 tests/App.test.js 12)
8477bd2e App.test.js       13)     it('has 1 child', () => {
5b512532 tests/App.test.js 14)         const tree = renderer.create(<App />).toJSON();
8477bd2e App.test.js       15)         expect(tree.children.length).toBe(1);
8477bd2e App.test.js       16)     });
5b512532 tests/App.test.js 17)
8477bd2e App.test.js       18)     it('renders correctly', () => {
5b512532 tests/App.test.js 19)         const tree = renderer.create(<App />).toJSON();
8477bd2e App.test.js       20)         expect(tree).toMatchSnapshot();
8477bd2e App.test.js       21)     });
8477bd2e App.test.js       22) });
```
</details>

We can use this to find a commit and review the contents. So the line which sets up the `PaperProvider` mock has this sha: 5b512532. We can review it with `git show`:

<details>
<summary>$ git show --oneline 5b512532</summary>

```
$ git show --oneline 5b512532
5b51253 (HEAD -> feature/create-habit-structure, origin/feature/create-habit-structure) Fix: Add mock paper provider
diff --git a/tests/App.test.js b/tests/App.test.js
index 1ed0a3f..5e00785 100644
--- a/tests/App.test.js
+++ b/tests/App.test.js
@@ -1,28 +1,22 @@
 import React from 'react';
-import { Provider } from 'react-redux';
 import renderer from 'react-test-renderer';
-import configureStore from 'redux-mock-store';
-
-const mockStore = configureStore([]);
-
 import App from '../App';

 jest.mock('redux-persist/integration/react', () => ({
     PersistGate: props => props.children,
 }));

+jest.mock('react-native-paper/lib/commonjs/core/Provider', () => 'PaperProvider');
+
 describe('<App />', () => {
-    let store;

-    beforeEach(() => {
-        store = mockStore({ count: { habits: {} } });
-    });
     it('has 1 child', () => {
-        const tree = renderer.create(<Provider store={store}><App /></Provider>).toJSON();
+        const tree = renderer.create(<App />).toJSON();
         expect(tree.children.length).toBe(1);
     });
+
     it('renders correctly', () => {
-        const tree = renderer.create(<Provider store={store}><App /></Provider>).toJSON();
+        const tree = renderer.create(<App />).toJSON();
         expect(tree).toMatchSnapshot();
     });
 });
\ No newline at end of file
```
</details>

We can review the git commit (in the example above "Fix: Add mock paper provider") and the changes in the commit to see what else happened around this line's change.

Here's a practical example:

- I was looking at the `Gemfile` of one of my projects (basically a `package.json` for Ruby) and saw [two gems that I usually don't install directly](https://github.com/exlibrisrpg/exlibris/blob/main/Gemfile#L49-L50) - `shoulda-context` and `shoulda-matchers` would usually be installed by `shoulda`
- With the "Blame" button in the top-right of the file viewer, I can see [an interactive version of the `git blame` output](https://github.com/exlibrisrpg/exlibris/blame/main/Gemfile#L49-L50) - they were both added in a commit titled "Order entries by name when they are listed"
- Clicking on that title shows me [the commit that changed those lines](https://github.com/exlibrisrpg/exlibris/commit/2ff8e3eac910a40dd2fe63c16fc09be01785e7d0) - from here I can see that I was fixing something relating to how items were sorted, and that I had to install the gems directly because [the `shoulda` gem was keeping me stuck on an older version of `shoulda-matchers`](https://github.com/exlibrisrpg/exlibris/commit/2ff8e3eac910a40dd2fe63c16fc09be01785e7d0#diff-89cade48462044ee1b672dc5f4c3ec250fbd29effcd8932096a23c1283c6731fL314-L316)
- I remember now that `shoulda-matchers` was having errors based on the way I was trying to sort things and those errors had been fixed in the newer version

It's not a perfect example because I probably should have explained that in the commit message - but it illustrates how keeping relevant changes together in individual commits can make it more useful to delve into the git history to understand how things got to the place they are.

So with that in mind, there are various ways we can take a PR which has a lot of commits and turn it into something that tells a clearer story. There's no harm in having loads of commits as you work - but tidying them up before merging can help you make the most of a git repo's history.

_However, if no-one else on a project is doing this then it's unlikely to get all that useful... I'll leave it as your call on whether you'd like to learn some of the ways to handle this._

File: src/components/CreateHabit.js
Comment (kmaaallen) [reply src/components/CreateHabit.js:12]: Have added a theme to the App - very basic for now but has some universal 'spacing' for use in styling as suggested. Can build on this as project progresses.

File: src/components/CreateHabit.js
Comment (kmaaallen) [reply src/components/CreateHabit.js]: Refactored component using hooks

File: src/components/LogButton.js
Comment (kmaaallen) [reply src/components/LogButton.js:72]: I hadn't thought of that - good catch. Have switched that now and also refactored this component to use hooks as well.

File: src/redux/actions.js
Comment (kmaaallen) [reply src/redux/actions.js]: I've updated the actions to use named parameters, I think that is the best approach as I imagine we will need to pass fuller payloads as the app functionality grows. I shall keep this in mind though if using positional args in future :)

File: src/redux/reducers.js
Comment (kmaaallen) [reply src/redux/reducers.js]: Agreed - updated

File: tests/LogButton.test.js
Comment (kmaaallen) [reply tests/LogButton.test.js:41]: Before each test I'm mocking the method so it returns a particular string (to get around the failing CI issues related to different locales), in line 33, just setting it back to as was before my mock. Is there a better way to do this? I couldn't find alternative way to get around the locale issue.

File: src/redux/reducers.js
Comment (kmaaallen) [reply src/redux/reducers.js]: Updated reducers to prevent mutating state. I've followed the documentation and think I've implemented this correctly now. Might be worth casting your eye over again just in case...

File: tests/CreateHabit.test.js
Comment (kmaaallen) [reply tests/CreateHabit.test.js:53]: @jdbann - Have been looking at this and implemented some more unit test using react-testing-library/native - allows us to test buttons etc. There are a few drawbacks - having some issues correctly identifying react native paper components and the unit tests aren't as streamlined as I would like. Might be something to look at in our next pairing session

**Comment by kmaaallen** [comment]
Have addressed most of the questions asked (let me know if anything missed / could be improved further). Would be interested to learn more about Git hygiene - shall I leave this PR open so we can use as an example or?

File: tests/CreateHabit.test.js
Comment (jdbann) [reply tests/CreateHabit.test.js:53]: Oh wow, this looks great! How does it feel as a test to you? Over-the-top or useful extra assurance? Do you feel inclined to add more of these as we go?

File: tests/LogButton.test.js
Comment (jdbann) [reply tests/LogButton.test.js:41]: It looks like `toLocaleString` takes a locale as an argument: `myDate.toLocaleString('en-GB')`

It might be that we have `en-GB` as our default locale but CI is in the US so has `en-US`?

**Comment by jdbann** [comment]
This is looking so great! I've replied to a couple of comments in here but feel free to merge whenever you fancy.

> Would be interested to learn more about Git hygiene - shall I leave this PR open so we can use as an example or?

That's up to you. With 70 commits this one might be a bit of a heavier one to start on - maybe we keep it in mind with the next one?

File: tests/CreateHabit.test.js
Comment (kmaaallen) [reply tests/CreateHabit.test.js:53]: I think this will be very useful for testing as we add more components on to the page. Some of the functionality feels like it is being tested twice - both here and in the reducers tests where we look at store changes. Might be worth streamlining these ones to just check a function is called rather that the outcome and the outcome is tested by the reducer. I'll keep working on it as we go.

File: tests/LogButton.test.js
Comment (kmaaallen) [reply tests/LogButton.test.js:41]: The toLocaleString is part of the code we want to test :
 Line 76  of LogButton component:  updatedDisplay: event ? (new Date(event.updated)).toLocaleString() : 'Never',
- Can't pass an en-US locale in the actual code as not all users will be in that region - hence the mocking. 
- I wouldn't have thought we could pass en-US in the mock either as then my local tests would fail?
- Maybe this is something we can pick up in our next session

**Comment by kmaaallen** [comment]
Makes sense - I'll merge this branch and we can start afresh talking about git hygiene etc.

**Pull request merged**
